### ISSUE TITLE
[Bug]: Screenshot reloads the page and resets viewport to default

### ISSUE THREAD

**Issue by mittster** [puppeteer/puppeteer#8690]
### Bug description

Taking a full page screenshot causes large page to reload. This in turn causes page to reset dimensions to 800x600 and throw away any lazy loaded content. Moreover, the width of screenshot is as expected (1366x1.5 in my case) and therefore it is not possible to detect if screenshot function failed and reloaded just by inspecting the size of screenshot. 
```

   const puppeteer = require('puppeteer');
   const imageSizeOf = require('image-size')   
   const browser = await puppeteer.launch({
        headless: true,
        defaultViewport: null,
        args: [
            '--window-size=1366,768',
            '--disable-dev-shm-usage',
            '--shm-size=3gb',
        ]
    });


    let page = await browser.newPage();
    await page.setViewport({ width: 1366, height: 768, deviceScaleFactor: 1.5 });
    let response = await page.goto("https://www.sinsay.com/si/sl/sale/woman/view-all-clothes", { waitUntil: "networkidle2" });
    await autoScroll(page); // not really necessary
    await page.evaluate(() => window.scrollTo(0, 0)); // not really necessary
    await page.screenshot({ path: desktopPath, fullPage: true });

    await browser.close();    


```

```
 async function autoScroll(page) {
    let results = await page.evaluate(async () => {
        try {
            return await new Promise((resolve, reject) => {
                let totalHeight = 0;
                let distance = 200;
                let counter = 0;
                let timer = setInterval(() => {
                    counter++;
                    var scrollHeight = document.body.scrollHeight;
                    window.scrollBy(0, distance);
                    totalHeight += distance;

                    if ((totalHeight >= scrollHeight - window.innerHeight) || counter > 80) {
                        clearInterval(timer);
                        resolve();
                    }
                }, 250);
            });

        } catch (e) {
            console.log("we got scrolling error:");
            console.log(e);

        }
    });
}
```

### Puppeteer version

14.3.0

### Node.js version

v16.15.0

### npm version

8.5.5

### What operating system are you seeing the problem on?

macOS

**Comment by mittster** [comment]
With additional testing I concluded that it is not reload, but resize event. Width jumps to 1 and then back to whatever was set before. The same problem is reported here: https://github.com/puppeteer/puppeteer/issues/7251

Adding ```captureBeyondViewport:false```is not a suitable solution because full page screenshots are going to miss a lot of website content.

 On same pages the screenshot looks ok. On others it breaks the page.  Is there a workaround?

**Comment by mittster** [comment]
This resize happens on all pages. You can track it with:
```
       // sets resize listener
        await page.evaluate((attributeName)=>{
            let meta = document.createElement("meta");
            let head = document.querySelector("head");
            head.append(meta);

            addEventListener('resize', (event) => {
                if(meta.hasAttribute(attributeName)){
                    let attribute = meta.getAttribute(attributeName);
                    meta.setAttribute(attributeName,attribute+"|"+window.innerWidth);
                }else{
                    meta.setAttribute(attributeName, ""+window.innerWidth);
                }
            });
        }, "mycustomattribute");
```

```
        // scereenshot
        await page.screenshot({path: "output.jpg",  fullPage: true});
```

 ```
       // sets 
        let resizeResult = await page.evaluate((attributeName) => {
            let meta = document.querySelector('meta['+attributeName+']');
            if(meta){
                let resizeValues =  meta.getAttribute(attributeName);
                return resizeValues;
            }
        }, "mycustomattribute");
        console.log(resizeResult); // this prints 1,1366 (1, previousWidth)
```


Even though resizing happens on all pages, screenshots on only some pages are broken. It appears that on some pages resize is fast enough and screenshot happens after the resizing has completed. 

This looks more and more like a chromium bug because Firefox doesn't have this issue. But it has many others so it isn't a viable alternative. 

Take a look at the video to see resizing in action:https://im.ezgif.com/tmp/ezgif-1-1426ad0f7e.gif

Is there an alternative for taking full page screenshots?

**Comment by OrKoN** [comment]
@mittster is the resize caused by the browser or is puppeteer actions cause the resize?

**Comment by mittster** [comment]
It's the action of taking screenshots.

**Comment by OrKoN** [comment]
I see, so I am able to observe the resize event but I don't see any negative effect on the screenshot. The screenshot of the full page works by resizing the viewport to the full height and then resizing it back to the original viewport. captureBeyondViewport flag is an alternative but you said it didn't work. It'd really help you if you could submit complete repro scripts with clear expected/actual results as it's difficult for people not familiar with the target website to judge if the screenshot looks okay or not. I saw you posted the video but I could not access it unfortunately.

You can also try to use the CDP method for capturing the screenshot directly and skip any logic that Puppeteer might be doing to see if some combination of parameters works better for your use case: https://chromedevtools.github.io/devtools-protocol/tot/Page/#method-captureScreenshot

**Comment by AnderssonPeter** [comment]
When I upgrade from puppeter 15 to 16 my screenshots get clipped incorrectly.
I just have `fullPage` set to `true`, I'm using puppeter with jest so its running multiple tabs in parallel all changing the viewport and taking screenshots.

I have tried to set `captureBeyondViewport` to `false` but that didn't fix the issue.

I'm not 100% sure its the same issue but I can't use puppeter 16 until its resolved :(

Something that confuses me what is the difference between `captureBeyondViewport` and `fullpage`?

**Comment by OrKoN** [comment]
@AnderssonPeter please file a separate issue with the repro script! `captureBeyondViewport` uses the CDP implementation to capture stuff that is outside the viewport (and might not be rendered) and `fullpage` uses the emulation to change the viewport to allow the content to be fully within the viewport and then takes the screenshot. Refer to the Page.ts code for the exact difference.

**Comment by AnderssonPeter** [comment]
@OrKoN I will try to create a repo that reproduces the issue.

**Comment by AnderssonPeter** [comment]
Sadly I wasn't able to find out exactly what causes this issue. and the project I'm working on is closed source so I can't give you that either..

**Comment by AnixPasBesoin** [comment]
Was anyone able to resolve this issue?

**Comment by OrKoN** [comment]
In general, the full-page screenshots in Chromium work by changing the viewport to match the dimensions of the page so a resize event is somewhat expected. The expectation is that the resize should not change anything on the page. Perhaps https://github.com/puppeteer/puppeteer/pull/8965 will help with this as there have been some CDP changes landed that might improve the situation.

**Comment by bradleysimard** [comment]
This problem happens even when fullPage is set to false, by the way.

It's definitely inconsistent. Sometimes my screenshot looks right, sometimes it seems to have taken the screenshot mid resize.

Still not sure why it's resizing when my viewport is already set.

**Comment by DeeHive** [comment]
Hello,
I tried with Puppeteer and then with 'chrome-remote-interface' to avoid any proper puppeteer logic and it is exactly the same behaviour. The resize logic sometimes reloads to page, sometimes make some elements disappear ...etc 

Is there a way to "deactivate" the page (all scripts ...etc) at least we are sure we take a screenshot of a specific state?

Thanks

**Comment by lawrenceching** [comment]
I'm working capturing full page of Grafana dashboard. And I'm totally blocked by this issue.
Grafana will recalculate the element position in order to dynamically adjust the panels on screen.

I saw people reported same issues to Puppeteer years ago but didn't get much attention.

I can capture the full page of Grafana dashboard manually, using Chrome extension like GoFullPage  or using Chrome devtool.
So I guess it is possible technically to fix it?

**Comment by gorshkov-leonid** [comment]
This problem is actual for me. We usee puppeteer for image snaphot testing via storybook. 
And the fact of resizing make puppeteer reload storybook's preview area which in turn fails tests.

**Comment by kamihork** [comment]
I had the same problem with pupeteer 22.13.0.

```
const puppeteer = require("puppeteer");

async function takeScreenshot() {
  const browser = await puppeteer.launch();
  const page = await browser.newPage();
  await page.goto("https://example.jp/", { waitUntil: "networkidle2" });
  await page.screenshot({ path: "screenshot.png", fullPage: true });
  await browser.close();
}

takeScreenshot();
```

**Comment by Bzzz** [comment]
> I'm working capturing full page of Grafana dashboard. And I'm totally blocked by this issue. Grafana will recalculate the element position in order to dynamically adjust the panels on screen.

I wasted the last couple of hours trying to capture a high-res screenshot of some Grafana Geomap and ChatGPT directed me onto the puppeteer route.
Grafana does weird things when loading the entire thing, Grafana needs extraordinary timeouts to finally and completely display items, Grafana doesn't work well with document.body.style.zoom, Grafana doesn't work well with supersized viewport widths/heights, and Grafana doesn't work well with deviceScaleFactor. Even when using pure Chrome and the device emulation mode that is capable of running viewports exceeding actual display size, Grafana will display things and then, upon capturing the screenshot, do one of the stupid reloads so that it just finishes in time to display headers and exactly nothing of the requested map. There's also a Grafana rendering plugin which cannot be installed the usual way, and that one has no ARM binaries so I would need to copy this whole mess onto a x86 machine for a lousy screenshot.

Let's face it: It's not an issue of puppeteer, Grafana is just a steaming pile of manure.
